src/features/utils/event-buffer.js

```diff
@@ -1,127 +1,80 @@
-  #held
+  #rawBytes = 0
+  #bufferBackup
+  #rawBytesBackup
 
   /**
-   *
-   * @param {number=} maxPayloadSize
+   * @param {number} maxPayloadSize
    */
   constructor (maxPayloadSize = MAX_PAYLOAD_SIZE) {
     this.maxPayloadSize = maxPayloadSize
   }
 
-  /**
-   * buffer is read only, use the helper methods to add or clear buffer data
-   */
-  get buffer () {
-    return this.#buffer
+  isEmpty () {
+    return this.#buffer.length === 0
   }
 
-  /**
-   * bytes is read only, use the helper methods to add or clear buffer data
-   */
-  get bytes () {
-    return this.#bytes
+  get () {
+    return this.#buffer
   }
 
-  /**
-   * held is another event buffer
-   */
-  get held () {
-    if (!this.#held) this.#held = new EventBuffer(this.maxPayloadSize)
-    return this.#held
+  byteSize () {
+    return this.#rawBytes
   }
 
-  /**
-   * Returns a boolean indicating whether the current size and buffer contain valid data
-   * @returns {boolean}
-   */
-  get hasData () {
-    return this.buffer.length > 0 && this.bytes > 0
+  wouldExceedMaxSize (incomingSize) {
+    return this.#rawBytes + incomingSize > this.maxPayloadSize
   }
 
   /**
-   * Adds an event object to the buffer while tallying size. Only adds the event if it is valid
-   * and would not make the event buffer exceed the maxPayloadSize.
-   * @param {Object} event the event object to add to the buffer
-   * @returns {EventBuffer} returns the event buffer for chaining
+   * Add feature-processed event to our buffer. If this event would cause our total raw size to exceed the set max payload size, it is dropped.
+   * @param {any} event - any primitive type or object
+   * @returns {Boolean} true if successfully added; false otherwise
    */
   add (event) {
-    const size = stringify(event).length
-    if (!this.canMerge(size)) return this
+    const addSize = stringify(event)?.length || 0 // (estimate) # of bytes a directly stringified event it would take to send
+    if (this.#rawBytes + addSize > this.maxPayloadSize) return false
     this.#buffer.push(event)
-    this.#bytes += size
-    return this
+    this.#rawBytes += addSize
+    return true
   }
 
   /**
-   * clear the buffer data
-   * @returns {EventBuffer}
+   * Wipes the main buffer
    */
   clear () {
-    this.#bytes = 0
     this.#buffer = []
-    return this
-  }
-
-  /**
-   * Hold the buffer data in a new (child) EventBuffer (.held) to unblock the main buffer.
-   * This action clears the main buffer
-   * @returns {EventBuffer}
-   */
-  hold () {
-    this.held.merge(this)
-    this.clear()
-    return this
+    this.#rawBytes = 0
   }
 
   /**
-   * Prepend the held EventBuffer (.held) back into the main buffer
-   * This action clears the held buffer
-   * @returns {EventBuffer}
+   * Backup the buffered data and clear the main buffer
+   * @returns {Array} the events being backed up
    */
-  unhold () {
-    this.merge(this.held, true)
-    this.held.clear()
-    return this
+  save () {
```

Comment:
Possibly missing a call to `this.clear()`

Reply:
save should not clear in the rewrite. Caller should call clear after save if such is the intention

Reply:
Ah, I see!  There's a shift in how the buffer works 👌